doc/model.rst

```diff
@@ -0,0 +1,719 @@
+
+When the model is started via ``Experiment.start()``, the TF function will be loaded to the
+standard orchestrator that is launched prior to the start of the model.
+
+-------------------
+TorchScript Scripts
+-------------------
+SmartSim supports the execution of TorchScripts
+with Models. Regardless of whether
+the orchestrator is colocated or not, each script added to the Model is
+loaded into the orchestrator prior to the execution of entity.
+Users have the flexibility to choose
+between `"GPU"` or `"CPU"` for device selection, and in environments with multiple
+devices, specific device numbers can be specified via `devices_per_node` such as `"GPU:1,"`.
+
+When specifying a TorchScript using ``Model.add_script()``, the
+following arguments are offered:
+
+- name (str): key to store script under.
+- script (t.Optional[str] = None): TorchScript code (only supported for non-colocated orchestrators).
+- script_path (t.Optional[str] = None): path to TorchScript code.
+- device (t.Literal["CPU", "GPU"] = "CPU"): device for script execution, defaults to “CPU”.
+- devices_per_node (int = 1): The number of GPU devices available on the host. This parameter only applies to GPU devices and will be ignored if device is specified as CPU.
+- first_device (int = 0): The first GPU device to use on the host. This parameter only applies to GPU devices and will be ignored if device is specified as CPU.
+
+You might use TorchScript scripts to represent individual models within the model.
+
+Example: Loading a TorchScript to the Model
+--------------------------------------------
+In this example, we demonstrate how to instruct SmartSim to load
+an TorchScript into the database at model runtime.
+
+**Python Script: Using Torch Scripts As ML Models**
+
+Define the TorchScript code as a variable in the Python driver script:
+
+.. code-block:: python
+
+    torch_script_str = "def negate(x):\n\treturn torch.neg(x)\n"
+
+**SmartSim Model Integration:**
+
+Assuming an initialized ``Model`` named `smartsim_model` exists, we add a TensorFlow model using
+the ``Model.add_script()`` function:
+
+.. code-block:: python
+
+    smartsim_model.add_script(name="example_script", script=torch_script_str, device="GPU", devices_per_node=2, first_device=0)
+
+In the above ``smartsim_model.add_script()`` code snippet, we offer the following arguments:
+
+-  name ("example_script"): key to store script under.
+-  script (torch_script_str): TorchScript code.
+-  device ("CPU"): device for script execution.
+-  devices_per_node (2): Use two GPUs per node.
+-  first_device (0): Start with 0 index GPU.
+
+When the model is started via ``Experiment.start()``, the TorchScript will be loaded to the
+orchestrator that is launched prior to the start of the model.
+
+=========================
+Data Collision Prevention
+=========================
+--------
+Overview
+--------
+SmartSim and SmartRedis can avoid key collision by prepending program-unique
```

Comment:
Following up on Chris' comment, maybe it is better to reword this paragraph as:

When multiple SmartSim models or ensemble members use the same name to reference a tensor, model, script, or other in-memory database object, collisions can occur that result in unintentional overwriting and accessing of data.  SmartSim assists users in avoiding these by automatically manipulating names through a process referred to as `key prefixing`.  When `key prefixing` is activated for an entity, names used to by the entity to reference in-memory database objects are atomically manipulated to avoid name collisions.

On the ``Model API``, there are several functions for activating, disabling, or checking `key prefixing` behavior on outgoing data actions (e.g. sending a tensor):

* Model.enable_key_prefixing()
* Model.disable_key_prefixing()
* Model.query_key_prefixing()

Additionally, if a user would like a ``Model`` to be able to access data from entities that are utilizing key-prefixing, then the user must add that entity (including the entity itself) to an internal inventory via ``Model.register_incoming_entity()``.   This inventory is used to configure the runtime environment of the SmartSim client (SmartRedis) to make visible other entity prefixing behavior.